### ISSUE TITLE
Event handlers style across ContainerJS

### ISSUE THREAD

**Issue by kirilpopov** [symphonyoss/ContainerJS#157]
Currently Electron (not sure about Open Fin ) uses event emitter style events - a typical examples 

```
window.on('resize', (event)=> { // handle});
```

Another alternative is to have specific methods per event 

```
window.onResize((bounds, window) => {});
```

My question is which way are we going to use in Container JS (It should be consistent across code base)?

I will try to list the pros and cons of two styles:

**Event emitter**

Pros

* new events: adding new events is easy, no real API change

Cons

* docs: documentation is hard (we will need to keep a list of events somewhere and figure out a way of documenting them)
* implementation in containers : will have to make sure every container implements the full set of events.

**Specific event methods**

Pros

* docs: auto documentation based on method comments
* implementation in containers: forced by the typescript interface 

Cons

* new events: adding new event requires API change

We must also consider end user usage, although this is based on experience and personal view; IMO usage with specific event methods is a bit better, because: 
1) different event callbacks can have different top level arguments
2) in typed languages (e.g. typescript) I won't need to depend on strings

(My vote is for specific event methods;)

**Comment by ColinEberhardt** [comment]
That's a very good question - thanks @kirilpopov 

One other aim of the Symphony APIs is that they should follow existing standards and conventions. DOM events make use of the `EventTarget` interface which has an [addEventListener](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) method which follows a similar pattern to the electron `on` method.

Another advantage of this approach (at least within the context of DOM), is extensibility, custom events can be added, events can also bubble.

I agree, documentation is an issue, but this is something that TypeScript will help with to a certain extent.

**Comment by BenLambertNcl** [comment]
@kirilpopov At the moment, the window has `addListener` and `removeListener` which work exactly the same as `on()` so changing to this would be very simple. I agree that not depending on strings would be nice, maybe we could have the option to use either strings or an enum of events, giving us useful type information without breaking what Electron/OpenFin already have?

**Comment by lspiro-Tick42** [comment]
In my opinion, adding both is not a good idea. It doubles the testing and documentation but biggest issue for me is that Developers should be able to understand the style of an API so they know how to use it. 
I know this doesn't help decide between Electron compatibility versus a more technically better (in some ways)  API

**Comment by kirilpopov** [comment]
Note that  Electron Windows are event emitters ( https://nodejs.org/api/events.html ) - this adds several really useful options for the end API users on top of EventTarget, e.g. using once to subscribe for event just once

```
window.once('eventName', ()=>{});
```

IMO if we're going with the string approach Windows (and other components that provide events) should be EventEmmiters (note that event emitters also implement EventTarget interface )

**Comment by ColinEberhardt** [comment]
> if we're going with the string approach Windows (and other components that provide events) should be EventEmmiters (note that event emitters also implement EventTarget interface )

That's a very good idea.

**Comment by BenLambertNcl** [comment]
Closed by #222
